src/CabanaPD_Neighbor.hpp

```diff
@@ -0,0 +1,219 @@
+    list_type _neigh_list;
+    double mesh_max[3];
+    double mesh_min[3];
+
+    Timer _timer;
+
+  public:
+    // Primary constructor: use positions and construct neighbors.
+    template <typename ParticleType, typename ModelType>
+    Neighbor( const bool half_neigh, const ModelType& model,
+              const ParticleType& particles, const double tol = 1e-14 )
+        : _half_neigh( half_neigh )
+    {
+        _timer.start();
+        for ( int d = 0; d < particles.dim; d++ )
+        {
+            mesh_min[d] = particles.ghost_mesh_lo[d];
+            mesh_max[d] = particles.ghost_mesh_hi[d];
+        }
+
+        if constexpr ( !is_contact<ModelType>::value )
+            _neigh_list =
+                list_type( particles.sliceReferencePosition(),
+                           particles.frozenOffset(), particles.localOffset(),
+                           model.cutoff() + tol, 1.0, mesh_min, mesh_max );
+        else
+            _neigh_list =
+                list_type( particles.sliceCurrentPosition(),
+                           particles.frozenOffset(), particles.localOffset(),
+                           model.cutoff() + tol, 1.0, mesh_min, mesh_max );
+        _timer.stop();
+    }
+
+    unsigned getMaxLocal()
+    {
+        auto neigh = _neigh_list;
+        unsigned local_max_neighbors;
+        auto neigh_max = KOKKOS_LAMBDA( const int, unsigned& max_n )
+        {
+            max_n = Cabana::NeighborList<list_type>::maxNeighbor( neigh );
+        };
+        Kokkos::RangePolicy<exec_space> policy( 0, 1 );
+        Kokkos::parallel_reduce( policy, neigh_max, local_max_neighbors );
+        Kokkos::fence();
+        return local_max_neighbors;
+    }
+
+    void getStatistics( unsigned& max_neighbors,
+                        unsigned long long& total_neighbors )
+    {
+        auto neigh = _neigh_list;
+        unsigned local_max_neighbors;
+        unsigned long long local_total_neighbors;
+        auto neigh_stats = KOKKOS_LAMBDA( const int, unsigned& max_n,
+                                          unsigned long long& total_n )
+        {
+            max_n = Cabana::NeighborList<list_type>::maxNeighbor( neigh );
+            total_n = Cabana::NeighborList<list_type>::totalNeighbor( neigh );
+        };
+        Kokkos::RangePolicy<exec_space> policy( 0, 1 );
+        Kokkos::parallel_reduce( policy, neigh_stats, local_max_neighbors,
+                                 local_total_neighbors );
+        Kokkos::fence();
+        MPI_Reduce( &local_max_neighbors, &max_neighbors, 1, MPI_UNSIGNED,
+                    MPI_MAX, 0, MPI_COMM_WORLD );
+        MPI_Reduce( &local_total_neighbors, &total_neighbors, 1,
+                    MPI_UNSIGNED_LONG_LONG, MPI_SUM, 0, MPI_COMM_WORLD );
+    }
+
+    const auto& list() const { return _neigh_list; }
+
+    // Only rebuild neighbor list as needed.
+    template <class ParticleType>
+    void update( const ParticleType& particles, const double search_radius,
```

Comment:
Consider doing `if constexpr` here as well for a `ModelType` check to protect against non-contact model updates.

Reply:
Done

Reply:
Looks good.